bin/holdings_deleter.rb

```diff
@@ -0,0 +1,132 @@
+# frozen_string_literal: true
+
+require "services"
+require "cluster"
+require "date"
+
+Services.mongo!
+
+STRING_FIELDS = [
+  "organization",
+  "local_id",
+  "enum_chron",
+  "n_enum",
+  "n_chron",
+  "status",
+  "condition",
+  "mono_multi_serial",
+  "country_code",
+  "uuid",
+  "issn"
+].freeze
+
+DATE_FIELDS  = ["date_received"].freeze
+INT_FIELDS   = ["ocn"].freeze
+FLOAT_FIELDS = ["weight"].freeze
+BOOL_FIELDS  = ["gov_doc_flag"].freeze
+CTRL_FIELDS  = ["noop", "verbose", "leave_empties"].freeze
+
+# Takes simple pairs of key=>value and constructs a update_many...pull query
+# based on those criteria.
+# Example:
+# bundle exec ruby bin/delete_matching_holdings.rb organization foo date_received 2020-01-01
+class HoldingsDeleter
+  attr_reader :matching_criteria
+
+  def initialize(args)
+    @matching_criteria = {}
+
+    Hash[*args].each do |key, val|
+      val_r = recast_val(key, val)
+      unless val_r.nil?
+        @matching_criteria[key] = val_r
+      end
+    end
+
+    if @matching_criteria.empty?
+      raise "No matching criteria given."
+    end
+
+    putsif "Criteria:"
+    @matching_criteria.each do |k, v|
+      putsif "#{k}:\t#{v} (#{v.class})"
+    end
+  end
+
+  def run
+    if @noop
+      puts "noop!"
+      return nil
+    end
+
+    result = Cluster.collection.update_many(
+      {},
+      { "$pull": { "holdings": { "$and": [@matching_criteria] } } }
+    )
+
+    delete_empty_clusters unless @leave_empties
+
+    putsif result.inspect
+    result
+  end
+
+  # This should perhaps be a method on Cluster, in some form?
+  # Definitely want extra eyeballs here making sure it's not wildly over-deleting.
+  def delete_empty_clusters
+    Cluster.collection.aggregate(
+      [
+        { "$match": { "ht_items.0":        { "$exists": 0 } } },
+        { "$match": { "holdings.0":        { "$exists": 0 } } },
+        { "$match": { "commitments.0":     { "$exists": 0 } } },
+        { "$match": { "ocn_resolutions.0": { "$exists": 0 } } }
+      ]
+    ).each do |empty_cluster|
+      putsif "Deleting empty cluster: #{empty_cluster}"
+      Cluster.find_by(ocns: empty_cluster["ocns"]).delete
```

Comment:
`Cluster.where({"$and":[{"ht_items.0": {"$exists": 0}},

                                      {"holdings.0":{"$exists":0}},

                                      {"commitments.0":{"$exists":0}},

                                     {"ocn_resolutions.0":{"$exists":0}}]

                                 }).each do |cluster|

  cluster.delete

end`

Reply:
that's the stuff